pkg/resources/kafka/configmap.go

```diff
@@ -37,17 +41,38 @@ func (r *Reconciler) configMapPod(broker banzaicloudv1alpha1.BrokerConfig, loadB
 			fmt.Sprintf("cruise.control.metrics.reporter.bootstrap.servers=%s\n", strings.Join(getInternalListeners(r.KafkaCluster.Spec.ListenersConfig.InternalListeners, broker, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled), ",")) +
 			fmt.Sprintf("broker.id=%d\n", broker.Id) +
 			generateStorageConfig(broker.StorageConfigs) +
-			generateAdvertisedListenerConfig(broker, r.KafkaCluster.Spec.ListenersConfig, loadBalancerIP, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled) +
+			generateAdvertisedListenerConfig(r.Client, broker, r.KafkaCluster.Spec.ListenersConfig, loadBalancerIP, r.KafkaCluster.Namespace, r.KafkaCluster.Name, r.KafkaCluster.Spec.HeadlessServiceEnabled) +
 			broker.Config},
 	}
 }
 
-func generateAdvertisedListenerConfig(broker banzaicloudv1alpha1.BrokerConfig, l banzaicloudv1alpha1.ListenersConfig, loadBalancerIP, namespace, crName string, headlessServiceEnabled bool) string {
+func generateAdvertisedListenerConfig(cli client.Client, broker banzaicloudv1alpha1.BrokerConfig, l banzaicloudv1alpha1.ListenersConfig, loadBalancerIP, namespace, crName string, headlessServiceEnabled bool) string {
 	advertisedListenerConfig := []string{}
-	for _, eListener := range l.ExternalListeners {
-		advertisedListenerConfig = append(advertisedListenerConfig,
-			fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), loadBalancerIP, eListener.ExternalStartingPort+broker.Id))
+
+	if loadBalancerIP != "" {
+		for _, eListener := range l.ExternalListeners {
+			advertisedListenerConfig = append(advertisedListenerConfig,
+				fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), loadBalancerIP, eListener.ExternalStartingPort+broker.Id))
+		}
+	} else {
+		var svc corev1.Service
+		var pods corev1.PodList
+		if err := cli.List(context.TODO(), &pods,
+			client.MatchingLabels{"kafka_cr": crName, "brokerId": string(broker.Id)}); err != nil {
+			log.Error(err)
+		}
+		pod := pods.Items[0]
+		if err := cli.Get(context.TODO(), types.NamespacedName{Namespace: namespace,
+			Name: fmt.Sprintf("%s-%d-svc", crName, broker.Id)}, &svc); err != nil {
+			log.Error(err)
+		}
+
+		for _, eListener := range l.ExternalListeners {
+			advertisedListenerConfig = append(advertisedListenerConfig,
+				fmt.Sprintf("%s://%s:%d", strings.ToUpper(eListener.Name), pod.Status.HostIP, svc.Spec.Ports[0].NodePort))
+		}
```

Comment:
I decided to get fancy on my branch and try out using a go template for the broker configuration. I'm not really tied to it though and can throw away my changes there if we want. I just thought going forward it might look a little cleaner. Here it is so you don't have to go fishing - https://github.com/tinyzimmer/kafka-operator/blob/master/pkg/resources/kafka/configmap.go#L32 

But if we do want to keep it, I'll need to re-implement this I think because it won't merge cleanly.

Reply:
It' s OK. I can wait until your branch is merged and I will rebase.